### PR TITLE
[WIP] Using Tox

### PR DESCRIPTION
**Pull request by saimn**
This adds a tox config, with envs both for pytest and setup.py, and use it on Travis and CircleCI (docs). 
If run in a python with conda and tox-conda, then a conda env is used.
So this still relies on ci-helpers to install miniconda, but then the other features like pinning versions are not used. Discussion needed here. We could imagine running a script inside the conda env to do what is done currently by ci-helpers, but I'm not sure about what exactly is needed here (it would probably be great to simplify this!).

```
❯ tox -lv
using tox.ini: /home/simon/dev/astropy/tox.ini
using tox-3.5.3 from /home/simon/.pyenv/versions/3.7.0/lib/python3.7/site-packages/tox/__init__.py
default environments:
py35-pytest     -> run tests with pytest
py35-pytest-all -> run tests with pytest
py35-setup      -> run tests with python setup.py
py35-setup-all  -> run tests with python setup.py
py36-pytest     -> run tests with pytest
py36-pytest-all -> run tests with pytest
py36-setup      -> run tests with python setup.py
py36-setup-all  -> run tests with python setup.py
py37-pytest     -> run tests with pytest
py37-pytest-all -> run tests with pytest
py37-setup      -> run tests with python setup.py
py37-setup-all  -> run tests with python setup.py
docs            -> invoke sphinx-build to build the HTML docs
dev             -> run with numpy and matplotlib dev versions
flake8          -> [no description]
```

TODO:
- [x] Appveyor
- [ ] need more flexibility with tox-conda to use conda only on a given env ? currently everything is run with conda on Travis
- [ ] testing the installed version. Currently `usedevelop` is always used, I will come back to this later.
- [ ] doctests (how is this done currently?)

See also:  #8202

### CODE REVIEW

**Comment by pllim** [comment]
cc @drdavella

File: .travis.yml [hunk 1]
```diff
@@ -78,89 +50,91 @@ matrix:
...
-               INSTALL_CMD='python setup.py build_ext --inplace'
-               PIP_DEPENDENCIES='pytest-astropy'
-               TEST_CMD='pytest --open-files --doctest-rst'
-               PYTEST_VERSION=3.7
-          script:
-            - $INSTALL_CMD
-            - $TEST_CMD
+          env: PYTHON_VERSION=3.5
+               TOXENV=py35-pytest
+               TOXOPTS='--develop --force-dep \'pytest==3.7\' --force-dep \'numpy==1.13.3\''
```
Comment (astrofrog) [inline .travis.yml]: I wonder if there is a way to improve this kind of syntax here - can we have a more simple ``TOX_FORCE_DEPS='pytest=3.7 numpy=1.13.3'`` and then later on convert that to the series of ``--force-dep`` flags?

File: .travis.yml [hunk 2]
```diff
@@ -78,89 +50,91 @@ matrix:
...
 
         - os: linux
           stage: Initial tests
-          env: PYTHON_VERSION=3.7 CONDA_DEPENDENCIES=$CONDA_ALL_DEPENDENCIES
-               PIP_DEPENDENCIES=$ASDF_PIP_DEP
-               SETUP_CMD='test -a "--durations=50"'
+          # language: python
+          # python: 3.7
+          # dist: xenial
+          # sudo: true
```
Comment (astrofrog) [inline .travis.yml]: Remove the comments?

File: .travis.yml [hunk 3]
```diff
@@ -78,89 +50,91 @@ matrix:
...
+               TOXENV=py36-pytest
+               TOXOPTS='--pre'
                EVENT_TYPE='push pull_request cron'
 
         # Do a PEP8/pyflakes test with flake8
         - os: linux
           stage: Initial tests
-          env: MAIN_CMD="flake8 astropy --count $FLAKE8_OPT" SETUP_CMD=''
+          # language: python
+          # python: 3.6
```
Comment (astrofrog) [inline .travis.yml]: Remove the comments?

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+requires = tox-venv
+           setuptools >= 30.0.0
+
+[testenv]
+usedevelop = True
+whitelist_externals = which
+passenv = HOME LC_ALL LC_CTYPE
+description =
+    setup: run tests with python setup.py
+    pytest: run tests with pytest
```
Comment (astrofrog) [inline tox.ini]: Add a description for 'all' that says with all optional dependencies?

**Comment by drdavella** [comment]
Here's an unsolicited opinion: `tox` should not ever be used to run `setup.py test`. Instead, it should only ever call `pytest` directly. A big reason for using `tox` is that it already handles all of the packaging and environment creation that we try to do in `setup.py test`. If we buy into `tox`, we can potentially stop maintaining this functionality entirely. Maybe this answers part of @mhvk's question.

The other major motivation for `tox`, as I see it, is the ability for push-button reproduction of CI environments. Currently it can be very difficult to reproduce tests that fail on CI. But with `tox`, we can keep large parts of the test environment constant between CI and local runs, which should make failures much easier to reproduce locally.

Yet another motivation is that the use of `tox` potentially obsoletes large parts of `ci-helpers`, which could allow us to substantially reduce the maintenance burden there.

**TL;DR**
The main motivations for exploring tox are:
* allows for push-button reproduction of Python environments on CI and locally
* potentially obsoletes large and complicated parts of Astropy's test infrastructure
* potentially obsoletes large parts of `ci-helpers`
* Allows us to shed complicated and painful-to-maintain parts of our custom infrastructure in favor of a tool that is very widely used in the rest of the Python community

File: .travis.yml [hunk 5]
```diff
@@ -78,89 +50,91 @@ matrix:
...
-               CONDA_DEPENDENCIES="$CONDA_ALL_DEPENDENCIES clang"
-               PIP_DEPENDENCIES="jplephem $ASDF_PIP_DEP"
-               CCOMPILER=clang
-               EVENT_TYPE='cron'
+          env: PYTHON_VERSION=3.6
+               TOXENV=py36-setup-all
+               TOXARGS='--remote-data=astropy'
+               # EVENT_TYPE='cron'
+               # CONDA_DEPENDENCIES="$CONDA_ALL_DEPENDENCIES clang"
+               # CCOMPILER=clang
```
[the remark this answers is not in the document]
Comment (saimn) [reply .travis.yml]: I don't think it would work: I can pass CC to the tox env, but currently the CONDA_DEPENDENCIES are not installed in the tox-conda envs. There are still installed by ci-helpers in the root env though, which is useless. If we want to keep these customization of the test env by ci-helpers, we could split it in at least two functions: one to just install miniconda, and another that would be called in the tox env.

[the remark this answers is not in the document]
Comment (saimn) [reply .travis.yml]: Also why using conda's clang here ? osx is using clang by default isn't it ?

File: .travis.yml (hunk 1 above)
Comment (saimn) [reply .travis.yml]: Yes maybe, that would be nicer.

File: .travis.yml (hunk 2 above)
Comment (saimn) [reply .travis.yml]: I left those comments because I wanted to try to have a build without conda, but I don't know yet how to do that ;)

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+conda_deps=
+    Cython
+    jinja2
+    numpy
+commands =
+    python --version
+    which python
+    pip freeze -l
+    pytest: pytest {posargs}
+    setup: python setup.py test {posargs}
```
Comment (drdavella) [inline tox.ini]: I mention this in the comments, but I don't think that it makes sense to run `setup.py test` using `tox`. Basically this means that Astropy is creating another test environment within a `tox` environment, which, in my opinion, is exactly what we'd like to avoid. Maybe in the short term we want to maintain this as a sanity check, but our eventual goal should be to get rid of it.

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+[testenv:dev]
+description = run with numpy and matplotlib dev versions
+deps =
+    -rpip-requirements-dev
+commands =
+    pip install git+https://github.com/numpy/numpy.git#egg=numpy
+    pip install git+https://github.com/matplotlib/matplotlib.git#egg=matplotlib
```
Comment (drdavella) [inline tox.ini]: Would it make more sense to move these to the `deps` section?

**Comment by mhvk** [comment]
@drdavella - thanks for the explanation of the use! It would indeed seem good to reduce the extra code we now have to help us test things, etc., especially as so few of us understand how it works. In that respect, you made quite an improvement already by making it possible to run `pytest` directly. Ideally, the `setup.py` way of doing it just becomes a nice short-cut (nice in the sense that `setup.py --help` tells you what is possible).

**Comment by saimn** [comment]
@mhvk - there were many discussions on this recently so I did not take the time to explain more, sorry. The goal is really to simplify things, so don't be afraid ;). 
The first goal is to make it easy to run test in a reproducible environment, the same locally and on Travis. Without duplicating dependencies in various places, running tests could be done the same way on Travis, CircleCI, Azure etc, without relying on specific scripts to setup the environments.
The future goal would be to remove `setup.py test` as tox provides a way to run tests in an isolated environment.

**Comment by saimn** [comment]
@drdavella - about `setup.py test`, this was a suggestion from @astrofrog to avoid changing too many things at once, which I think is reasonable. There nothing wrong about running `setup.py test` inside tox I think, we still get the improvement on the dependencies installation. We also need to find a way to test the installed version, which `setup.py test` does.

File: tox.ini (hunk 7 above)
Comment (saimn) [reply tox.ini]: This would conflict with pip-requirements-dev I think.

**Comment by mhvk** [comment]
@saimn - OK, thanks. Please do link to those discussions at the top...

File: tox.ini (hunk 7 above)
Comment (drdavella) [reply tox.ini]: Okay. I understand the reason for using a `pip-requirements` file but I wonder if you could have finer-grained control over dependencies by just including them specifically in the `tox.ini` file. Then you could prevent the conflict by using factors.

**Comment by drdavella** [comment]
@saimn yes that's reasonable, and I agree. We should just remember that the eventual goal should be to remove this duplication. Eventually I think it might be reasonable for `setup.py test` to really just be an alias for `tox.main` that uses a particular standard environment.

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,68 @@
+[tox]
+envlist = py{35,36,37}-{pytest,setup}{,-all},docs,dev,flake8
+skipsdist = True
+requires = tox-venv
+           setuptools >= 30.0.0
+
+[testenv]
+usedevelop = True
+whitelist_externals = which
+passenv = HOME LC_ALL LC_CTYPE
```
Comment (Cadair) [inline tox.ini]: It's also worth passing `CC` here so that people who are using ccache etc can still use it with tox.

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,68 @@
+[tox]
+envlist = py{35,36,37}-{pytest,setup}{,-all},docs,dev,flake8
+skipsdist = True
+requires = tox-venv
+           setuptools >= 30.0.0
+
+[testenv]
+usedevelop = True
```
Comment (Cadair) [inline tox.ini]: Same comment as above about the `skipsdist` stuff?

**Comment by astrofrog** [comment]
Just to mention another benefit of tox: most of the configuration for CI then lives in a single file regardless of CI platform, which makes it much easier to switch between different CI services and also means that most users (once we do this in the package-template) will just need to worry about editing ``tox.ini`` rather than each individual CI config file when e.g. updating dependencies.

File: tox.ini (hunk 4 above)
Comment (saimn) [reply tox.ini]: Done.

File: tox.ini (hunk 8 above)
Comment (saimn) [reply tox.ini]: Yep, I added CC and CFLAGS, which should fix the coverage for C files.

**Comment by astrofrog** [comment]
@saimn - I've merged https://github.com/astropy/astropy/pull/8198, so could you see if you can use the extras_require instead?

**Comment by saimn** [comment]
Latest news: 
- updated with extra_requires, which works well, except for bintrees which doesn't compile on py37 with clang. I removed it as it was not in the requirements file, I think it was deprecated in favor of sortedcontainers.
- `--force-dep` works only for packages listed in `deps` (https://github.com/tox-dev/tox/issues/534, wontfix), we could do with it but I'm trying to replace it with environment variables, which would also also to specify numpy version for the conda package.

**Comment by saimn** [comment]
The pytest 3.7 builds are failing with doctestplus, is that a known issue ? cc @drdavella 
```
______________________________ ERROR collecting  _______________________________
/home/travis/build/saimn/astropy/.tox/py35-setup-all/lib/python3.5/site-packages/_pytest/runner.py:201: in __init__
    self.result = func()
/home/travis/build/saimn/astropy/.tox/py35-setup-all/lib/python3.5/site-packages/_pytest/runner.py:261: in <lambda>
    call = CallInfo(lambda: list(collector.collect()), "collect")
/home/travis/build/saimn/astropy/.tox/py35-setup-all/lib/python3.5/site-packages/_pytest/main.py:475: in collect
    for x in self._collect(arg):
/home/travis/build/saimn/astropy/.tox/py35-setup-all/lib/python3.5/site-packages/_pytest/main.py:519: in _collect
    for x in root._collectfile(pkginit):
E   AttributeError: 'DocTestModulePlus' object has no attribute '_collectfile'
```

**Comment by saimn** [comment]
Ok, so there is now a build with Travis' native Python (no conda), to show that we can use both in parallel. And the flake8 build also use this. The remaining question for me is, what is vision for this and ci-helpers. Currently ci-helpers always setup a miniconda and installs a few things, which is useless with tox since it recreates a conda env.

With this PR we can have builds with miniconda and/or native Python, setup.py and/or pytest, optionally pin version for numpy, pytest, or whatever. Are we missing something from ci-helpers, other than the possibility to control dependencies for many packages from one place ? (Which I'm not sure if this is really a feature).

About skipsdist/usedevelop (cc @Cadair ): the normal use of tox is to create the sdist, install it in a virtualenv, and run the tests. Currently I skipped the sdist and use the develop mode because this is useless when running tests with `setup.py test`, and it takes some time. But if we switch to pytest as the main way to proceed, then we will want to be sure to test the installed package.

**Comment by astrofrog** [comment]
> With this PR we can have builds with miniconda and/or native Python, setup.py and/or pytest, optionally pin version for numpy, pytest, or whatever. Are we missing something from ci-helpers, other than the possibility to control dependencies for many packages from one place ? (Which I'm not sure if this is really a feature).

Indeed the main benefit of ci-helpers is to control dependencies/environment variables from one place. The ability to pin some packages from a centralized location or set global environment variables is really useful - for instance recently build with Matplotlib have started to segfault because of a new version of PyQt5, and @bsipocz fixed it by simply setting ``MPLBACKEND=Agg`` in ci-helpers. So I think that in the long term we'll still want to source a file from ci-helpers, but that code can become simpler and simpler. This doesn't mean we have to continue to use conda (that is, there could be a mode in ci-helpers that is more geared to native python + conda) though I do think we want to keep a few conda builds since that's what a lot of users use. We could potentially have a mode that just installs miniconda and an environment with the right Python version but nothing else.

**Comment by bsipocz** [comment]
@saimn - Yes, as @astrofrog  already mentioned the real aim of `ci-helpers` is to have one central place to add the workarounds for upstream issues. The latest one was indeed matplotlib, but pytest/sphinx/conda pinnings are regular there. As well as to ensure/force the latest astropy being picked up for the tests, though that I believe was a conda specific issue so should not need a workaround.

Also over the years, ci-helpers grew to be this monster to work around workarounds, so if we go back to those basics, it would be a welcomed change (while we need to keep the current monster around for compatibility, there are just way too many users relying on it).

File: tox.ini (hunk 6 above)
Comment (obestwalter) [reply tox.ini]: `setup.py test` is even actively discouraged in the docs: https://tox.readthedocs.io/en/latest/example/basic.html#integration-with-setup-py-test-command

File: tox.ini [hunk 10]
```diff
@@ -0,0 +1,68 @@
+[tox]
+envlist = py{35,36,37}-{pytest,setup}{,-all},docs,dev,flake8
+skipsdist = True
```
[the remark this answers is not in the document]
Comment (obestwalter) [reply tox.ini]: This only means that no package should be built. That plays together with using `usedevelop = True` to test the project installed in develop mode (a.k.a. `pip install --editable <project path>`).

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (astrofrog) [reply .travis.yml]: Sorry I didn't reply to this sooner - I think the motivation for this might have been when we were considering adding some OpenMP functionality, but since we didn't do that in the end I think it's fine to drop this.

File: tox.ini [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    dev: git+https://github.com/numpy/numpy.git#egg=numpy
+    dev: git+https://github.com/matplotlib/matplotlib.git#egg=matplotlib
+extras =
+    test
+    all: all
+conda_deps=
+    Cython
+    jinja2
+    numpy{env:NUMPY_VERS:}
+    all: scipy
```
Comment (astrofrog) [inline tox.ini]: Why is this needed? There should be a wheel, so I would have thought it would work fine without?

**Comment by astrofrog** [comment]
@saimn - looking at this again and I think it would be great to try and polish this up. Here are a few thoughts on proceeding:

* I think the ``dev`` build should instead use the same syntax as other test builds (it's not obvious that it's a test command), so e.g. ``py35-pytest-dev``. That would also allow us to use the same modes/Python versions as for other test options.

* Do things go wrong if we were to make use of ``PYTEST_VERSION`` and ``NUMPY_VERSION`` instead of ``PYTEST_VERS`` and ``NUMPY_VERS``? At the moment it's confusing to have two separate sets of variables.

* I think we need a dedicated page in the dev docs to explain how to get started with using tox.
Having good documentation explaining how to get set up with and run tox is going to be essential to convincing some people to approve these changes. Maybe you could start something here which doesn't have to be extensive but gives the motivation and instructions?

* What is the difference between ``TOXARGS`` and ``TOXOPTS``?

**Comment by astrofrog** [comment]
Just out of curiosity, is there a way that the environment name could include pinnings such as numpy, e.g ``tox -e py37-pytest-np1.14`` rather than use environment variables, or does that explode the complexity of the tox.ini file?

**Comment by astrofrog** [comment]
Thinking about this a little more, one of the things that bothers me a bit is the fact that we still have some of the customization/config in e.g. the travis file which means we don't have quite a 'one-click' solution for reproducing CI builds locally. Therefore I was wondering whether one option would be that in addition to the standard environments you've included, we also have a few environments named ``ci-...`` that reproduce some of the specific customizations done in the CI. For example, we could have ``ci-legacy`` that runs the tests with the oldest compatible versions of all dependencies and Python version, ``ci-remote`` which runs the tests with the full remote data activated, and the idea is that these could be much more specialized configurations but would avoid having to configure them through the Travis CI file, so that they are not CI platform dependent.

File: tox.ini (hunk 11 above)
Comment (saimn) [reply tox.ini]: You mean for scipy, or all deps here ? 
My idea was to install at least a few common packages from conda instead of pip.

Comment (astrofrog) [reply tox.ini]: Ah ok - I think it would be good to have a comment to explain this

**Comment by saimn** [comment]
- about `dev`: it can done with the same syntax as other envs, but it means more envs. And the `pyXX` is actually not needed in most case, you can already use the `dev` env with any python version. The `pyXX` is only useful if you have several Python versions installed.

- `NUMPY_VERSION` vs `NUMPY_VERS`: I used a different name because for me it was more confusing to use the same name to do different things. Also `NUMPY_VERSION` means that the conda base envs would spend some useless time to install another numpy package. But in the end, `ci-helpers` is used to install miniconda, and is currently doing some stuff for nothing, so if we want to go further with tox we should think about the future of ci-helpers.

- docs: yes sure, I don't have much time currently (and also have some wip for fits), but I agree that it will be needed.

- `TOXARGS` is for arguments passed to the test command (pytest or `setup.py test`), and `TOXOPTS` is for tox arguments (`--pre`, `--develop`).

- About env with specific dependencies (Numpy) and ci builds reproduction: it's probably a matter of taste, but I don't think that the tox config must contain all the customisations for the CI builds. I mean in principle it would be fine, but I fear that this will add a lot of complexity to the tox config. And there are many possible options to support, so you will need one tox env per CI build, but this will be difficult to do with the factorization features. I don't know which option is better on the long term, but if you want to go that way it needs some work.

File: tox.ini (hunk 11 above)
Comment (saimn) [reply tox.ini]: Yep, done.

**Comment by saimn** [comment]
I rebased and added the `pyXX-...-dev` envs.

File: tox.ini [hunk 12]
```diff
@@ -0,0 +1,67 @@
+[tox]
+envlist = py{35,36,37}-{pytest,setup}{,-all,-dev},docs,flake8
+skipsdist = True
+requires = tox-venv
+           setuptools >= 30.0.0
```
Comment (astrofrog) [inline tox.ini]: Minor: this should actually be 30.3.0

File: tox.ini [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+[testenv]
+usedevelop = True
+whitelist_externals = which
+passenv = HOME LC_ALL LC_CTYPE CC CFLAGS
+description =
+    pytest: run tests with pytest,
+    setup: run tests with python setup.py,
+    all: and all dependencies,
+    dev: and numpy and matplotlib dev versions,
+# declaring numpy and pytest is needed for --force-dep
```
Comment (astrofrog) [inline tox.ini]: Just to check, is this still needed since --force-dep isn't used?

File: tox.ini (hunk 12 above)
Comment (saimn) [reply tox.ini]: Fixed, thanks.

File: tox.ini (hunk 13 above)
Comment (saimn) [reply tox.ini]: Yes indeed, I removed this comment.

**Comment by saimn** [comment]
Tests are failing because of a Numpy multiarray failure on Windows, and pytest_plugins on Travis:

> Defining 'pytest_plugins' in a non-top-level conftest is no longer supported because it affects the entire directory tree in a non-explicit way.
  /home/travis/build/astropy/astropy/astropy/conftest.py
Please move it to a top level conftest file at the rootdir:
  /home/travis/build/astropy/astropy

I hope that  I did not introduce regression with the rebase..
@drdavella - any idea about the pytest_plugins issue ?

**Comment by astrofrog** [comment]
Let's plan to re-open this once/if the APE on the future astropy-helpers is approved.

**Comment by astrofrog** [comment]
@saimn - I've squashed and included the commits from here in #9726 (and kept you as an author)
